core/state/statedb.go

```diff
@@ -66,6 +70,34 @@ func (n *proofList) Delete(key []byte) error {
 	panic("not supported")
 }
 
+// CheckTransactionPredicates checks the stateful precompile predicates of any of the given
+// transactions that reference a stateful precompile address in their access list.
+// Returns [len(txs), nil] if and only if all referenced predicates are met.
+// Otherwise, returns the index of the first transaction that was invalid and the predicate error.
+func CheckTransactionPredicates(rules params.Rules, snowCtx *snow.Context, txs types.Transactions, proposerVMBlockCtx *block.Context) (int, error) {
+	precompileConfigs := rules.Precompiles
+	for i, tx := range txs {
+		for _, accessTuple := range tx.AccessList() {
+			precompileConfig, isPrecompileAccess := precompileConfigs[accessTuple.Address]
+			if !isPrecompileAccess {
+				continue
+			}
+
+			predicate := precompileConfig.Predicate()
+			if predicate == nil {
+				continue
+			}
+
+			if err := predicate(snowCtx, proposerVMBlockCtx, utils.HashSliceToBytes(accessTuple.StorageKeys)); err != nil {
+				log.Debug("Transaction predicate verification failed.", "txId", tx.Hash(), "precompileAddress", accessTuple.Address.Hex())
+				return i, err
```

Comment:
My understanding is that we only need to return the index of the first transaction that has a predicate error as the caller will remove the failing transaction and all of the remaining transactions from the transaction pool. We want to remove all of the remaining transactions because, although not explicitly specified in the interface, the transactions are actually from the same origin account and sorted by the nonce. 

It might be helpful to add a comment to explain this property of the transactions array. Otherwise, it is not clear why can't process the next transaction after a failed predicate execution.

Reply:
Added comments to clarify these points.